sound/soc/sof/ipc4-mtrace.c

```diff
@@ -0,0 +1,607 @@
+		return;
+
+	msg.primary = SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG);
+	msg.primary |= SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST);
+	msg.primary |= SOF_IPC4_MOD_ID(SOF_IPC4_MOD_INIT_BASEFW_MOD_ID);
+	msg.primary |= SOF_IPC4_MOD_INSTANCE(SOF_IPC4_MOD_INIT_BASEFW_INSTANCE_ID);
+	msg.extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_FW_PARAM_ENABLE_LOGS);
+
+	priv->state_info.enable = 0;
+
+	msg.data_size = sizeof(priv->state_info);
+	msg.data_ptr = &priv->state_info;
+	iops->set_get_data(sdev, &msg, msg.data_size, true);
+
+	priv->mtrace_state = SOF_MTRACE_DISABLED;
+
+	for (i = 0; i < sdev->num_cores; i++) {
+		struct sof_mtrace_core_data *core_data = &priv->cores[i];
+
+		core_data->host_read_ptr = 0;
+		core_data->dsp_write_ptr = 0;
+		wake_up(&core_data->trace_sleep);
+	}
+}
+
+/*
+ * Parse the slot descriptors at debug_box offset, we are only interested in the
+ * type of the slot to map it to a core
+ */
+static void sof_mtrace_find_core_slots(struct snd_sof_dev *sdev)
+{
+	struct sof_mtrace_priv *priv = sdev->fw_trace_data;
+	struct sof_mtrace_core_data *core_data;
+	u32 slot_desc_offset, type, core;
+	int i;
+
+	/*
+	 * Out of the 16 (MAX_MTRACE_SLOTS) slots the first one is allocated for
+	 * descriptors to describe the remaining 15 slots (MAX_MTRACE_SLOTS - 1)
+	 * We are checking the descriptors to find the slot <-> core mapping
+	 */
+	slot_desc_offset = sdev->debug_box.offset;
+	for (i = 0; i < MAX_MTRACE_SLOTS - 1; i++) {
+		/* Read the slot type */
+		sof_mailbox_read(sdev, slot_desc_offset + 4, &type, 4);
+		if ((type & SOF_MTRACE_SLOT_TYPE_MASK) == SOF_MTRACE_SLOT_DEBUG_LOG) {
+			core = type & SOF_MTRACE_SLOT_CORE_MASK;
+
+			if (core >= sdev->num_cores)
+				continue;
+
+			core_data = &priv->cores[core];
+			core_data->slot_offset = sdev->debug_box.offset;
+			core_data->slot_offset += SOF_MTRACE_SLOT_SIZE * (i + 1);
+			dev_dbg(sdev->dev, "%s: slot%d is used for core%u\n",
+				__func__, i, core);
+			if (core_data->missed_update) {
+				sof_ipc4_mtrace_update_pos(sdev, core);
+				core_data->missed_update = false;
+			}
+		} else if (type) {
+			dev_dbg(sdev->dev, "%s: slot%d is not a log slot (%#x)\n",
+				__func__, i, type);
+		}
+
+		slot_desc_offset += sizeof(u32) * 3;
+	}
+}
+
+static int ipc4_mtrace_init(struct snd_sof_dev *sdev)
+{
+	struct sof_ipc4_fw_data *ipc4_data = sdev->private;
+	struct sof_mtrace_priv *priv;
+	int i, ret;
+
+	if (sdev->fw_trace_data) {
+		dev_err(sdev->dev, "fw_trace_data has been already allocated\n");
+		return -EBUSY;
+	}
+
+	if (!ipc4_data->mtrace_log_bytes ||
+	    ipc4_data->mtrace_type != SOF_IPC4_MTRACE_INTEL_CAVS2X) {
+		sdev->fw_trace_is_supported = false;
+		return 0;
+	}
+
+	priv = devm_kzalloc(sdev->dev, struct_size(priv, cores, sdev->num_cores),
+			    GFP_KERNEL);
+	if (!priv)
+		return -ENOMEM;
+
+	sdev->fw_trace_data = priv;
+
+	/* default enable default trace state_info */
+	priv->state_info.aging_timer_period = 10;
+	priv->state_info.fifo_full_timer_period = 10;
```

Comment:
magic numbers, and units not specified.

Reply:
OK, I'll add something `*_DEFAULT_TIMER_PERIOD 10` for these and try to find the unit, I can not recall seeing it, but it should be somewhere.